pkg/restic/repository_manager.go

```diff
@@ -0,0 +1,574 @@
+	// CheckRepo checks the specified repo for errors.
+	CheckRepo(name string) error
+
+	// CheckAllRepos checks all repos for errors.
+	CheckAllRepos() error
+
+	// PruneRepo deletes unused data from a repo.
+	PruneRepo(name string) error
+
+	// PruneAllRepos deletes unused data from all
+	// repos.
+	PruneAllRepos() error
+
+	// Forget removes a snapshot from the list of
+	// availabel snapshots in a repo.
+	Forget(repo, snapshotID string) error
+
+	Backupper
+	Restorer
+}
+
+// Backupper can execute restic backups of volumes in a pod.
+type Backupper interface {
+	// BackupPodVolumes backs up all annotated volumes in a pod.
+	BackupPodVolumes(ctx context.Context, backup *arkv1api.Backup, pod *corev1api.Pod, log logrus.FieldLogger) error
+}
+
+// Restorer can execute restic restores of volumes in a pod.
+type Restorer interface {
+	// RestorePodVolumes restores all annotated volumes in a pod.
+	RestorePodVolumes(ctx context.Context, restore *arkv1api.Restore, pod *corev1api.Pod, log logrus.FieldLogger) error
+}
+
+type BackendType string
+
+const (
+	AWSBackend   BackendType = "aws"
+	AzureBackend BackendType = "azure"
+	GCPBackend   BackendType = "gcp"
+
+	credsSecret = "restic-credentials"
+)
+
+type repositoryManager struct {
+	objectStore   cloudprovider.ObjectStore
+	bucket        string
+	arkClient     clientset.Interface
+	secretsClient corev1client.SecretInterface
+	log           logrus.FieldLogger
+	repoPrefix    string
+	repoLocker    *repoLocker
+}
+
+// NewRepositoryManager constructs a RepositoryManager.
+func NewRepositoryManager(
+	objectStore cloudprovider.ObjectStore,
+	config arkv1api.ObjectStorageProviderConfig,
+	arkClient clientset.Interface,
+	secretsClient corev1client.SecretInterface,
+	log logrus.FieldLogger,
+) RepositoryManager {
+	rm := &repositoryManager{
+		objectStore:   objectStore,
+		bucket:        config.ResticLocation,
+		arkClient:     arkClient,
+		secretsClient: secretsClient,
+		log:           log,
+		repoLocker:    newRepoLocker(),
+	}
+
+	switch BackendType(config.Name) {
+	case AWSBackend:
+		url := "s3.amazonaws.com"
+
+		// non-AWS, S3-compatible object store
+		if config.Config != nil && config.Config["s3Url"] != "" {
+			url = config.Config["s3Url"]
+		}
+
+		rm.repoPrefix = fmt.Sprintf("s3:%s/%s", url, rm.bucket)
+	case AzureBackend:
+		rm.repoPrefix = "azure:" + rm.bucket + ":"
+	case GCPBackend:
+		rm.repoPrefix = "gs:" + rm.bucket + ":"
+	}
+
+	return rm
+}
+
+func (rm *repositoryManager) BackupPodVolumes(ctx context.Context, backup *arkv1api.Backup, pod *corev1api.Pod, log logrus.FieldLogger) error {
+	// get volumes to backup from pod's annotations
+	volumesToBackup := GetVolumesToBackup(pod)
+	if len(volumesToBackup) == 0 {
+		return nil
+	}
+
+	// ensure a repo exists for the pod's namespace
+	if err := rm.ensureRepo(pod.Namespace); err != nil {
+		return err
+	}
+
+	resultChan := make(chan *arkv1api.PodVolumeBackup)
+
+	informer := arkv1informers.NewFilteredPodVolumeBackupInformer(
```

Comment:
We will probably want to have 1 informer running per backup, and this does 1 informer per backup+pod. We could add a new struct that would be per-backup, and create a new one at the beginning of a backup.